### PR TITLE
chore: convert codebase to TS

### PR DESCRIPTION
**Pull request by ankeetmaini**
Please provide enough information so that others can review your pull request:

## Motivation

Explain the **motivation** for making this change. What existing problem does the pull request solve?

As per #202

## Test plan

Demonstrate the code is solid and make sure you add tests to cover the new functionality. 

The code must pass tests.

## Code formatting

Look around. Match the style of the rest of the codebase. Run `yarn format` before committing.

- [x] use Typescript with strict flags enabled right from the beginning, use react-native-typescript-transformer instead of Babel
- [x] convert the source to TS with types
- [x] remove flow reference
- [x] add `ts-jest` to run tests after being converted to TS
- [x] use typescript-eslint for linting and disable corresponding prettier settings for typescript:recommended rules
- [x] remove all anys as much as possible
- [x] port example app to TS
- [x] pass all failing tests (4 failing tests as of today)
- [x] add bundler - rollup?
- [x] fix skipped saga tests

### CODE REVIEW

File: src/components/NetworkContext.tsx [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { createContext } from 'react';
+
+export type ConnectivityState = {
```
Comment (ddzirt) [inline src/components/NetworkContext.tsx]: Why do we need this? We have 
```javascript
export type NetworkState = {
  isConnected: boolean;
  actionQueue: EnqueuedAction[];
};
```

if we do something like:
```javascript
interface ConnectivityState {
isConnected: boolean,
}

interface NetworkState extends ConnectivityState {
  actionQueue: EnqueuedAction[]
}
```

we would be able to just reuse typing in helpers and components instead of creating this one. Also currently `react-native-offline` is not utilizing the latest version of `react-native-netifo` because in this file `src/components/NetworkConnectivity.tsx` there are deprication warnings.

File: src/utils/checkConnectivityInterval.ts [hunk 2]
```diff
@@ -1,6 +1,5 @@
-// @flow
+let interval: number | null = null;
```
Comment (ddzirt) [inline src/utils/checkConnectivityInterval.ts]: why do we need `number | null`?

`let interval: number = null` this has to be enough.

File: src/utils/makeHttpRequest.ts [hunk 3]
```diff
@@ -71,8 +70,11 @@ export default function makeHttpRequest({
           status: this.status,
         });
       };
-      Object.keys(headers).forEach((key: string) => {
-        xhr.setRequestHeader(key, headers[key]);
+      
+      
+      Object.keys(headers).forEach(key => {
+        const k = key as keyof typeof headers;
```
Comment (ddzirt) [inline src/utils/makeHttpRequest.ts]: and this change would not be required due to proper typing for `headers`

File: src/utils/makeHttpRequest.ts [hunk 4]
```diff
@@ -23,10 +21,11 @@ type ResolvedValue = {
   status: number,
 };
 
+const CACHE_HEADER_VALUE = 'no-cache, no-store, must-revalidate';
```
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply src/utils/makeHttpRequest.ts]: I thought the headers need to be limited to just these three values and hence I didn't introduce index signature. We can do this too, then anything can be passed if that's what we want!

[the remark this answers is not in the document]
Comment (ddzirt) [reply src/utils/makeHttpRequest.ts]: OK, well if we do leave this as a constant then I think it should be placed within `utils/constants.ts`

File: src/utils/checkConnectivityInterval.ts (hunk 2 above)
Comment (ankeetmaini) [reply src/utils/checkConnectivityInterval.ts]: This doesn't let me assign null as null is not of type number.

Comment (ddzirt) [reply src/utils/checkConnectivityInterval.ts]: strictNullChecks=true. Ok

**Comment by ankeetmaini** [comment]
Update:

- I've now fixed all type errors in the `src` folder except files related to `Sagas`. I'll start on it shortly after reading up on them once.

- I've not tried to run tests which I'll once I finish `saga` files too.

- Once all test cases pass, I'll add ESLint which will catch more bugs and unintentional usage of `any`s if any.

- The example app too is remaining to be typed

**Comment by rgommezz** [comment]
Thank you very much @ankeetmaini for your work so far! 🚀 I'll try to take a look soon at your progress :)

File: .vscode/launch.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (ankeetmaini) [inline .vscode/launch.json]: delete this

File: .vscode/settings.json [hunk 6]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (ankeetmaini) [inline .vscode/settings.json]: delete this

**Comment by ankeetmaini** [comment]
Update: I've added eslint-typescript to the project. There are 24 lint errors and 49 TS errors, mostly in the example folder, which is yet to be typed.

I've had to add `skipLibCheck: true` as globals from react-native and `dom.lib.d.ts` were colliding giving a lot of errors.

**Comment by ankeetmaini** [comment]
@rgommezz this is passing the tests now, everything is green, such a happy new year :D

I've to run the example app to see if everything works correctly, which I'll this week. But this is open for review now. :)

File: src/redux/createReducer.ts [hunk 7]
```diff
@@ -34,29 +36,26 @@ function handleOfflineAction(
...
       typeof actionToLookUp === 'object'
-        ? { ...actionToLookUp, meta }
+        ? ({ ...actionToLookUp, meta } as FluxAction)
         : actionToLookUp;
     const similarActionQueued = comparisonFn(
       actionWithMetaData,
       state.actionQueue,
     );
-
+    const actions = similarActionQueued
```
Comment (rgommezz) [inline src/redux/createReducer.ts]: Rename to `newActionQueue`

File: src/redux/createReducer.ts [hunk 8]
```diff
@@ -1,27 +1,29 @@
...
 import { SEMAPHORE_COLOR } from '../utils/constants';
-import actionTypes from './actionTypes';
 import getSimilarActionInQueue from '../utils/getSimilarActionInQueue';
-import type {
-  FluxAction,
-  FluxActionWithPreviousIntent,
-  FluxActionForRemoval,
+import {
   NetworkState,
+  nonNullable,
```
Comment (rgommezz) [inline src/redux/createReducer.ts]: `nonNullable` is a function, not a type. Let's move to `utils` folder

File: src/redux/createReducer.ts [hunk 9]
```diff
@@ -94,11 +95,13 @@ function handleChangeQueueSemaphore(
   };
 }
 
-export default (comparisonFn: Function = getSimilarActionInQueue) => (
-  state: NetworkState = {
-    ...initialState,
-  },
-  action: *,
+type ComparisonFn = (
```
Comment (rgommezz) [inline src/redux/createReducer.ts]: Could we move this type definition to the top? It's also used by the function above.

File: src/redux/sagas.ts [hunk 10]
```diff
@@ -1,38 +1,35 @@
...
 import { put, select, call, take, cancelled, fork } from 'redux-saga/effects';
-import { eventChannel } from 'redux-saga';
+import { eventChannel, Subscribe } from 'redux-saga';
 import { AppState, Platform } from 'react-native';
 import NetInfo from '@react-native-community/netinfo';
 import { networkSelector } from './createReducer';
 import checkInternetAccess from '../utils/checkInternetAccess';
 import { connectionChange } from './actionCreators';
-import type { HTTPMethod } from '../types';
+import DEFAULT_ARGS from '../utils/defaultConnectivityArgs';
```
Comment (rgommezz) [inline src/redux/sagas.ts]: I just realised this could be added to the `constants.ts` file, so that all the defaults are compiled into one single place.

File: src/types.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  pingTimeout: number;
+  pingServerUrl: string;
+  shouldPing: boolean;
+  pingInterval: number;
+  pingOnlyIfOffline: boolean;
+  pingInBackground: boolean;
+  httpMethod: HTTPMethod;
+};
+
+export type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;
```
Comment (rgommezz) [inline src/types.ts]: I think `Omit` is a built in utility type

File: src/types.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+// https://stackoverflow.com/a/58110124/2615091
+export function nonNullable<T>(value: T): value is NonNullable<T> {
```
Comment (rgommezz) [inline src/types.ts]: This function shouldn't be here.

File: src/utils/defaultConnectivityArgs.ts [hunk 13]
```diff
@@ -0,0 +1,18 @@
+import { ConnectivityArgs } from '../types';
```
Comment (rgommezz) [inline src/utils/defaultConnectivityArgs.ts]: Let's better move all this into constants.ts and make it a named export

File: src/utils/getSimilarActionInQueue.ts [hunk 14]
```diff
@@ -1,5 +1,6 @@
-// @flow
+// TODO import directly to reduce bundle-size?
 import { isEqual } from 'lodash';
```
Comment (rgommezz) [inline src/utils/getSimilarActionInQueue.ts]: Yes, let's do `import isEqual from 'lodash/isEqual' for potential tree shaking.

File: src/utils/objectEntries.ts [hunk 15]
```diff
@@ -0,0 +1,5 @@
+const entries = Object.entries as <T>(
```
Comment (rgommezz) [inline src/utils/objectEntries.ts]: What do we need this file for?

File: test/createNetworkMiddleware.test.ts [hunk 16]
```diff
@@ -181,29 +183,29 @@ describe('createNetworkMiddleware with different REGEX config', () => {
...
+  const fetchMockData = (dispatch: Dispatch) =>
     new Promise(resolve => {
       setTimeout(() => {
         dispatch({ type: 'FETCH_DATA_SUCCESS' });
         resolve();
       }, 1000);
     });
 
-  function fetchData(dispatch) {
+  const fetchData = (dispatch: Thunk) => {
```
Comment (rgommezz) [inline test/createNetworkMiddleware.test.ts]: dispatch: Dispatch?

File: test/reducer.test.ts [hunk 17]
```diff
@@ -148,30 +149,28 @@ describe('OFFLINE_ACTION action type', () => {
...
         const prevState = getState(false, thunkFactory(1, 'Bilbo', 55));
-        const thunk = actionCreators.fetchOfflineMode(
-          thunkFactory(2, 'Link', 54),
-        );
+        const thunk = thunkFactory(2, 'Link', 54);
+        const wrappedThunk = actionCreators.fetchOfflineMode(thunk);
 
         expect(getSimilarActionInQueue(thunk, prevState.actionQueue)).toEqual(
-          prevState.actionQueue[0].action,
+          prevState.actionQueue[0],
```
Comment (rgommezz) [inline test/reducer.test.ts]: Why this change and the test still passes? 🤔

File: test/reducer.test.ts [hunk 18]
```diff
@@ -394,15 +395,19 @@ describe('dismiss feature', () => {
 
 describe('networkSelector', () => {
   it('returns the correct shape', () => {
-    const state = {
+    const state: { network: NetworkState } = {
       network: {
         isConnected: true,
-        actionQueue: ['foo'],
+        // actionQueue contains `EnqueuedAction` objects? Correct?
```
Comment (rgommezz) [inline test/reducer.test.ts]: Correct, good catch! You can remove the comment :)

File: test/sagas.test.ts [hunk 19]
```diff
@@ -14,27 +18,17 @@ import networkSaga, {
...
-  pingOnlyIfOffline: false,
-  pingInBackground: false,
-  httpMethod: DEFAULT_HTTP_METHOD,
-};
+const args = DEFAULT_ARGS;
 
 describe('sagas', () => {
   describe('networkSaga', () => {
-    it('forks netInfoChangeSaga with the right params', () => {
+    it.skip('forks netInfoChangeSaga with the right params', () => {
```
Comment (rgommezz) [inline test/sagas.test.ts]: What issues are you having with these tests?

**Comment by ankeetmaini** [comment]
Thanks @rgommezz for the review, will update this PR by the weekend! :)

File: example/DummyNetworkContext.tsx [hunk 20]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+type Context = {
```
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply example/DummyNetworkContext.tsx]: Removed, it was yelling I think in some older version of TS. No error now!

File: example/components/ActionButtons.tsx [hunk 21]
```diff
@@ -5,18 +5,22 @@ import {
...
+  subOneType,
+  otherType,
+  cancelOtherType,
+} from '../redux/actions';
 
-const Touchable = Platform.select({
-  android: TouchableNativeFeedback,
-  ios: TouchableOpacity,
-});
+const Touchable = TouchableHighlight;
```
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply example/components/ActionButtons.tsx]: Haha, I was getting a type error

```
Type 'typeof TouchableNativeFeedback' is not assignable to type 'typeof TouchableOpacity'.
  Type 'TouchableNativeFeedback' is missing the following properties from type 'TouchableOpacity': setOpacityTo, setTimeout, clearTimeout, setInterval, and 11 more.ts(2322)
```

File: example/components/ActionButtons.tsx [hunk 22]
```diff
@@ -48,13 +52,18 @@ const showInfo = actionType => () => {
     cancelable: false,
   });
 };
-
+type ActionTypes = {
```
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply example/components/ActionButtons.tsx]: done

File: example/components/TabBarIcon.tsx [hunk 23]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import { Icon } from 'expo';
+
+import Colors from '../constants/Colors';
+
+type Props = {
+  name: string;
+  focused: boolean;
+};
+export default class TabBarIcon extends React.PureComponent<Props> {
```
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply example/components/TabBarIcon.tsx]: made it a stateless function

File: example/screens/ComponentsScreen.tsx [hunk 24]
```diff
@@ -123,22 +123,22 @@ const styles = StyleSheet.create({
     marginRight: 8,
   },
   tabBarInfoContainer: {
-    position: 'absolute',
+    position: 'absolute' as 'absolute',
```
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply example/screens/ComponentsScreen.tsx]: removed this, let me see if it yells

File: package.json [hunk 25]
```diff
@@ -6,11 +6,11 @@
   "author": "Raul Gomez Acuña <raulgdeveloper@gmail.com> (https://github.com/rgommezz)",
   "license": "MIT",
   "scripts": {
-    "lint": "eslint src",
+    "lint": "eslint --ext .js,.ts,.tsx {src,test}/**/*.{ts,tsx}",
```
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply package.json]: I'll add a rollup config to transpile

File: src/components/NetworkConnectivity.tsx [hunk 26]
```diff
@@ -1,34 +1,21 @@
...
-} from '../utils/constants';
+import { ConnectivityArgs } from '../types';
+import { ConnectivityState } from './NetworkContext';
+import DEFAULT_ARGS from '../utils/defaultConnectivityArgs';
 
 export type RequiredProps = {
-  children: (state: State) => React.Node,
-};
+  children: (state: ConnectivityState) => React.ReactNode;
+} & Partial<DefaultProps>;
```
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply src/components/NetworkConnectivity.tsx]: Wow!

File: src/components/NetworkProvider.tsx [hunk 27]
```diff
@@ -0,0 +1,26 @@
...
+import { ConnectivityArgs } from '../types';
+import DEFAULT_ARGS from '../utils/defaultConnectivityArgs';
+
+type Props = {
+  children: React.ReactNode;
+} & Partial<ConnectivityArgs>;
+
+NetworkProvider.defaultProps = DEFAULT_ARGS;
+
+function NetworkProvider(props: Props) {
```
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply src/components/NetworkProvider.tsx]: done

```js
function NetworkProvider({
  httpMethod = DEFAULT_ARGS.httpMethod,
  pingInBackground = DEFAULT_ARGS.pingInBackground,
  pingInterval = DEFAULT_ARGS.pingInterval,
  pingOnlyIfOffline = DEFAULT_ARGS.pingOnlyIfOffline,
  pingServerUrl = DEFAULT_ARGS.pingServerUrl,
  pingTimeout = DEFAULT_ARGS.pingTimeout,
  shouldPing = DEFAULT_ARGS.shouldPing,
  ...rest
}: Props) {
  const { children } = rest;
```

File: src/components/ReduxNetworkProvider.tsx [hunk 28]
```diff
@@ -54,11 +62,9 @@ class ReduxNetworkProvider extends React.Component<Props> {
   }
 }
 
-function mapStateToProps(state: { network: NetworkState }) {
-  return {
-    isConnected: state.network.isConnected,
-  };
-}
+const mapStateToProps = (state: AppState, _?: OwnProps) => ({
```
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply src/components/ReduxNetworkProvider.tsx]: this is so that correct type is inferred for the connect component!

I think I struggled fair bit and generally it's much verbose but [this link](https://github.com/piotrwitek/react-redux-typescript-guide#connect-with-react-redux) is what I followed as far as I remember!

File: src/index.ts [hunk 29]
```diff
@@ -0,0 +1,20 @@
+import _NetworkProvider from './components/NetworkProvider';
```
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply src/index.ts]: yes, so stupid of me, sorry :D

File: src/redux/createNetworkMiddleware.ts [hunk 30]
```diff
@@ -0,0 +1,192 @@
...
+  regexActionType: RegExp,
+  actionTypes: ActionType,
+): boolean {
+  return (
+    isObjectAndShouldBeIntercepted(action, regexActionType, actionTypes) ||
+    isThunkAndShouldBeIntercepted(action)
+  );
+}
+
+function didComeBackOnline(action: AllActions, wasConnected: boolean) {
```
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply src/redux/createNetworkMiddleware.ts]: I don't think it'll work, `AllActions` is same as `EnqueuedAction` so I tried

```ts
export type EnqueuedAction = FluxAction | Thunk | {};
```

but got 

```
Property 'type' does not exist on type 'EnqueuedAction'.
  Property 'type' does not exist on type '{}'
```

File: src/redux/createNetworkMiddleware.ts [hunk 31]
```diff
@@ -0,0 +1,192 @@
...
+    return (
+      action.type === networkActionTypes.CHANGE_QUEUE_SEMAPHORE &&
+      isQueuePaused &&
+      action.payload === SEMAPHORE_COLOR.GREEN
+    );
+  }
+  return false;
+}
+
+type GetState = Pick<MiddlewareAPI<Dispatch, State>, 'getState'>['getState'];
```
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply src/redux/createNetworkMiddleware.ts]: Wow, great catch! thanks :)

File: src/utils/objectEntries.ts (hunk 15 above)
Comment (ankeetmaini) [reply src/utils/objectEntries.ts]: this is used in NetworkConnectivityTest

File: test/createNetworkMiddleware.test.ts (hunk 16 above)
Comment (ankeetmaini) [reply test/createNetworkMiddleware.test.ts]: oops, yes!

File: src/components/ReduxNetworkProvider.tsx (hunk 28 above)
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply src/components/ReduxNetworkProvider.tsx]: I've removed it, it still picks the correct defs!

File: package.json (hunk 25 above)
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply package.json]: done!

File: rn-cli.config.js [hunk 32]
```diff
@@ -0,0 +1,8 @@
+module.exports = {
```
[the remark this answers is not in the document]
Comment (ankeetmaini) [reply rn-cli.config.js]: I'll take a look!

**Comment by rgommezz** [comment]
@all-contributors please add @ankeetmaini for code

**Comment by ankeetmaini** [comment]
Yaaaay! Just woke up to this! Thanks so much @rgommezz! One of the best codebases I worked on. Learnt a lot! 🥳🥳🥳

**Comment by grit96** [comment]
Is there any possibility of adding a flow [libdef](https://flow.org/en/docs/libdefs/) so this library can still support flow?
